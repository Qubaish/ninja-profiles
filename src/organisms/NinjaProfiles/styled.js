import styled from 'styled-components';
import { device } from '../../utils/breakpoints';

export const StyledContainer = styled.div`
    width: 100%;
    margin-left: auto;
    margin-right: auto;
    @media ${device.md} { 
        width: 80%;
    }
`;

export const StyledProfileWrapper = styled.div`
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    position: relative;
    overflow: hidden;
    justify-content: space-around;
    .profile-card {
        margin-bottom: 1.8rem;
    }
`;
