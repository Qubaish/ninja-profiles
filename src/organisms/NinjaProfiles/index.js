import React, { useState, useEffect } from 'react';
import InfiniteScroll from 'react-infinite-scroller';
import ProfileCard from '../../molecules/ProfileCard';
import Filters from '../../atoms/Filters';
import { apiUrl, profilesToShow } from '../../utils/constant';
import { StyledContainer, StyledProfileWrapper } from './styled';

let globalArr = [];
let filteredArr = [];
let recordsCount = 0;

export default function NinjaProfiles() {

    const [profiles, setProfiles] = useState([]);
    const [hasMore, setHasMore] = useState(false);
    const [offices, setOffices] = useState([]);

    const getProfiles = () => {
        fetch(apiUrl)
        .then((resp) => resp.json())
        .then(function(data) {
            globalArr = sortProfilesByNameAndLocation(data);
            recordsCount = globalArr.length;
            checkHasMore(recordsCount);
            setOffices([...new Set(data.map(p => p.office))]);
            setProfiles([...globalArr].slice(0, profilesToShow));
        });
    }

    useEffect(() => {
        getProfiles();
    }, []);

    const sortProfilesByNameAndLocation = data => {
        const sortedProfiles = data.sort(
            function(a, b) {          
               if (a.office === b.office) {
                  return a.name.localeCompare(b.name);
               }
               return a.office > b.office ? 1 : -1;
        });
        return sortedProfiles;
    };

    const onNameChange = ({target}) => {
        const { value } = target;
        if (value && value.length > 1) {
            filteredArr = [...globalArr].filter(c => c.name.toLowerCase().includes(value.toLowerCase()));
            recordsCount = filteredArr.length;
            checkHasMore(recordsCount);
            setProfiles(filteredArr.slice(0, profilesToShow));
        };
        if(!value) {
            setProfilesToDefault();
        }
    }

    const onLocationSelect = ({target}) => {
        const { value } = target;
        if (value) {
            filteredArr = [...globalArr].filter(d => d.office === value);
            recordsCount = filteredArr.length;
            checkHasMore(recordsCount);
            setProfiles(filteredArr.slice(0, profilesToShow));
        }
        if(!value) {
            setProfilesToDefault();
        }
    }

    const checkHasMore = length => length >= profilesToShow ? setHasMore(true) : setHasMore(false);

    const setProfilesToDefault = () => {
        setProfiles(globalArr.slice(0, profilesToShow));
        setHasMore(true);
        filteredArr = [];
        recordsCount = globalArr.length;
    }

    const loadFunc = page => {
        const recordsToShow = page * profilesToShow;
        if( recordsToShow >= recordsCount) {
            setHasMore(false);
        }
        const arrToPush = filteredArr.length > 0 ? [...filteredArr] : [...globalArr];
        const profiles = arrToPush.slice(0, recordsToShow);
        setProfiles(profiles);
    };

    return (
        <StyledContainer>
            <Filters cities={offices} onNameChange={onNameChange} onLocationSelect={onLocationSelect}/>
            {profiles.length === 0 && (<h1>No Records Found</h1>)}
            <InfiniteScroll
                pageStart={1}
                loadMore={loadFunc}
                initialLoad={false}
                threshold={20}
                hasMore={hasMore}
                loader={
                <div className="loader" key={0}>
                    Loading ...
                </div>
                }
            >
                <StyledProfileWrapper data-testid="profile-wrapper">
                    {profiles.map((profile, i) => (
                        <ProfileCard className="profile-card" profile={profile} key={i}/>
                    ))}
                </StyledProfileWrapper>
            </InfiniteScroll>
        </StyledContainer>
    );
}