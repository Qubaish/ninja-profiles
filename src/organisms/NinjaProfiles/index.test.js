import React from 'react';
import { render, fireEvent, waitForElement } from '@testing-library/react';
import NinjaProfiles from './index';

const mockProfile = [{
    name: "Simon",
    office: "LUND",
}, {
    name: "Lint",
    office: "LUND",
}, {
    name: 'john',
    office: 'Stockholm'
}];


global.fetch = jest.fn(() =>
    Promise.resolve({
        json: () => Promise.resolve([]),
    })
);
describe('<NinjaProfiles />', () => {
    beforeEach(() => {
        fetch.mockClear();
    });
    it('checking snapshot', async () => {
        const { asFragment } = render(<NinjaProfiles />);
        const elem = await waitForElement(() => asFragment());
        expect(elem).toMatchSnapshot();
    });
    it('checking records not found', async() => {
        const { getByText } = render(<NinjaProfiles />);
        const elem = await waitForElement(() => getByText(/No Records Found/i));
        expect(elem).toBeInTheDocument();
    });
    it('render profiles with correct quantity', async () => {
        global.fetch = jest.fn(() =>
            Promise.resolve({
                json: () => Promise.resolve([...mockProfile, ...mockProfile]),
            })
        );
        const { getByTestId } = render(<NinjaProfiles />);
        const elem = await waitForElement(() => getByTestId('profile-wrapper'));
        expect(elem.children).toHaveLength(6);
    });
    it('search profiles with incorrect name', async () => {
        global.fetch = jest.fn(() =>
            Promise.resolve({
                json: () => Promise.resolve([...mockProfile]),
            })
        );
        const { getByTestId, getByText } = render(<NinjaProfiles />);
        const elem = await waitForElement(() => getByTestId('input'));
        fireEvent.change(elem, { target: { value: 'SimonMock' } })
        const msg = await waitForElement(() => getByText(/No Records Found/i));
        expect(msg).toBeInTheDocument();
    });
    it('search profiles with correct name', async () => {
        global.fetch = jest.fn(() =>
            Promise.resolve({
                json: () => Promise.resolve([...mockProfile]),
            })
        );
        const { getByTestId, getByText } = render(<NinjaProfiles />);
        const elem = await waitForElement(() => getByTestId('input'));
        fireEvent.change(elem, { target: { value: 'Lint' } })
        const msg = await waitForElement(() => getByText(/Lint/i));
        expect(msg).toBeInTheDocument();
    });
    it('search profiles with correct offic location', async () => {
        global.fetch = jest.fn(() =>
            Promise.resolve({
                json: () => Promise.resolve([...mockProfile]),
            })
        );
        const { getByTestId } = render(<NinjaProfiles />);
        const elem = await waitForElement(() => getByTestId('select'));
        fireEvent.change(elem, { target: { value: 'LUND' } })
        const profiles = await waitForElement(() => getByTestId('profile-wrapper'));
        expect(profiles.children).toHaveLength(2);
    });
    it('search profiles with empty office location', async () => {
        global.fetch = jest.fn(() =>
            Promise.resolve({
                json: () => Promise.resolve([...mockProfile]),
            })
        );
        const { getByTestId } = render(<NinjaProfiles />);
        const elem = await waitForElement(() => getByTestId('select'));
        fireEvent.change(elem, { target: { value: "" } })
        const profiles = await waitForElement(() => getByTestId('profile-wrapper'));
        expect(profiles.children).toHaveLength(3);
    });
});