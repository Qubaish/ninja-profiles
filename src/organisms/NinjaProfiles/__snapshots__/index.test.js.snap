// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NinjaProfiles /> checking snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-AxmLO ebeJGn"
  >
    <div
      class="sc-AxgMl cpXvYN"
    >
      <input
        class="sc-AxheI hSpogh"
        data-testid="input"
        name="name"
        placeholder="Name e.g John"
      />
      <select
        class="sc-Axmtr fGszkM"
        data-testid="select"
        name="city"
      >
        <option
          value=""
        >
          Filter by Offices
        </option>
      </select>
    </div>
    <h1>
      No Records Found
    </h1>
    <div>
      <div
        class="sc-fzozJi jWrzyX"
        data-testid="profile-wrapper"
      />
    </div>
  </div>
</DocumentFragment>
`;
